plugins/out_s3/s3.c

```diff
@@ -964,45 +958,38 @@ static int cb_s3_init(struct flb_output_instance *ins,
         ctx->timer_ms = UPLOAD_TIMER_MIN_WAIT;
     }
 
-    /* 
-     * S3 must ALWAYS use sync mode 
-     * In the timer thread we do a mk_list_foreach_safe on the queue of uplaods and chunks
-     * Iterating over those lists is not concurrent safe. If a flush call ran at the same time
-     * And deleted an item from the list, this could cause a crash/corruption. 
-     */
-    ctx->s3_client->upstream->flags &= ~(FLB_IO_ASYNC);
-
-    /* clean up any old buffers found on startup */
-    if (ctx->has_old_buffers == FLB_TRUE) {
-        flb_plg_info(ctx->ins,
-                     "Sending locally buffered data from previous "
-                     "executions to S3; buffer=%s",
-                     ctx->fs->root_path);
-        ctx->has_old_buffers = FLB_FALSE;
-        ret = put_all_chunks(ctx, FLB_TRUE);
-        if (ret < 0) {
-            ctx->has_old_buffers = FLB_TRUE;
-            flb_plg_error(ctx->ins,
-                          "Failed to send locally buffered data left over "
-                          "from previous executions; will retry. Buffer=%s",
-                          ctx->fs->root_path);
-        }
-    }
-
-    /* clean up any old uploads found on start up */
-    if (ctx->has_old_uploads == FLB_TRUE) {
-        flb_plg_info(ctx->ins,
-                     "Completing multipart uploads from previous "
-                     "executions to S3; buffer=%s",
-                     ctx->stream_upload->path);
-        ctx->has_old_uploads = FLB_FALSE;
-
-        /*
-         * we don't need to worry if this fails; it will retry each
-         * time the upload callback is called
-         */
-         cb_s3_upload(config, ctx);
-    }
+    /* this would use sync IO which we want to avoid */
```

Comment:
I suppose that this means old data will not be sent until the first chunk is received. Maybe it's fine to use sync networking here since it's only once. We just don't want to consistantly use sync network.

If so, cb_s3_upload may need a new parameter called is_async

Reply:
Yea I'm still debating this part myself. And testing. This is one bit which can block shutdown from happening properly.

Reply:
This is just temporary while I'm testing. I need to get the flush startup chunks routine working and testing in the daemon coroutine first, then I'll add this back.

Currently the flush startup chunks routine is flushing all chunks, since it just iterates over the fstore queue...


We have to have this because otherwise S3 won't upload any old data until its sent new data, which is silly.